vignettes/custom-functions.Rmd

```diff
@@ -55,45 +56,91 @@ This is the reason why it is better to use polars expressions in your own functi
+There is `<expr>$map_batches()` in this package, but it is not recommended to use it because it is unstable and may not work as expected in the streaming mode.
+
+Instead, it is recommended to convert the polars object to an R DataFrame or similar, and then use common higher-order functions like `lapply()` or `purrr::map()`.
 
 Note that when you convert your polars object to R, you may run out of memory if your object is a LazyFrame that is too big to fit in memory once collected.
 Another crucial point is that we want the R function to run as fast as possible on the data.
-One way to do that is to take advantage of purrr to run the function [in parallel](https://purrr.tidyverse.org/reference/in_parallel.html).
+One way to do that is to take advantage of `{purrr}` to run the function [in parallel](https://purrr.tidyverse.org/reference/in_parallel.html).
 
-```{r, eval=check_purrr_version}
+```{r}
 library(purrr)
+```
+
+If we do not want to run the function in parallel, something like this will suffice:
 
+```{r}
+dat |>
+  as.data.frame() |>
+  map(\(col) scale(col)[, 1]) |>
+  as_polars_df()
+```
+
+When running in parallel, we need to create background processes using `mirai::daemons()`
+(see the documentation of `{purrr}` for details).
+
+```{r, eval=purrr_parallel_available}
 # Set the number of background processes to use
 mirai::daemons(3)
+```
 
+```{r, eval=purrr_parallel_available}
+#| echo: false
+# Ensure load the package to the daemons
+mirai::everywhere({
+  if (!rlang::is_installed("polars")) pkgload::load_all()
+})
+```
+
+After starting the daemons, we can wrap the function passed to `purrr::map()`
+with `purrr::in_parallel()` to run it in parallel in the background processes.
+
+```{r, eval=purrr_parallel_available}
 dat |>
   as.data.frame() |>
-  map(in_parallel(\(s) scale(s)[, 1])) |>
-  _[] |>
+  map(in_parallel(\(col) scale(col)[, 1])) |>
   as_polars_df()
 ```
 
-Note that this applies the function on the entire DataFrame (or LazyFrame).
-To apply this to a subset of columns only, we can use `map_at()`:
+Note that this applies the function on the entire data on memory and all data is collected in R.
+To apply this to a subset of columns only, we can use `purrr::map_at()`.
+
+In this case, instead of using `as.data.frame()` to convert all data to R,
+we can use `as.list()` to convert the polars DataFrame to a list of Series
+and only convert some Series to R, thus avoiding unnecessary conversions between
+Polars and R.
 
-```{r, eval=check_purrr_version}
+In technical terms, polars Series and other objects are wrappers around external pointers,
+which usually cannot be sent to other processes.
+However, this package has built-in integration with `{mirai}`,
+which registers the serialization and deserialization functions for Series and other objects
+when the package is loaded.
+This allows users to treat Series and other polars objects like regular R objects.
```

Comment:
This should be a footnote, it's not necessary to know for average users and breaks the reading flow.

Reply:
I don't know if it is not a need-to-know for the average users, but I agree that it is worth moving to footnotes.

Reply:
@etiennebacher altdoc with mkdocs does not seem supporting footnotes?
<img width="1195" height="194" alt="image" src="https://github.com/user-attachments/assets/361471dc-a275-41da-9084-685c21b6533d" />

Reply:
I'll take a look